vsb/databases/opensearch/opensearch.py

```diff
@@ -0,0 +1,207 @@
+
+        # Bulk ingest documents
+        upload_response = self.client.bulk(body=actions)
+
+    def update_batch(self, batch: list[Record]):
+        self.insert_batch(batch)
+
+    def search(self, request: SearchRequest) -> list[str]:
+        @retry(
+            wait=wait_exponential_jitter(initial=0.1, jitter=0.1),
+            stop=stop_after_attempt(5),
+            after=after_log(logger, logging.DEBUG),
+        )
+        def do_query_with_retry():
+            query = {
+                "size": request.top_k,
+                "fields": ["vsb_vec_id"],
+                "_source": False,
+                "query": {
+                    "knn": {
+                        "v_content": {"vector": request.values, "k": self.dimensions}
+                    }
+                },
+            }
+            return self.client.search(body=query, index=self.index_name)
+
+        response = do_query_with_retry()
+        # sending the VSB Id's of the top k results
+        vsb_id = []
+        [vsb_id.append(m["fields"]["vsb_vec_id"][0]) for m in response["hits"]["hits"]]
+        return vsb_id
+
+    def fetch_batch(self, request: list[str]) -> list[Record]:
+        # Fetching records not directly supported; requires implementation of metadata storage
+        raise NotImplementedError("fetch_batch not supported for OpenSearch")
+
+    def delete_batch(self, request: list[str]):
+        # deleting the records not directly supported; requires implementation of delete by vsb_vec_id in the metadata
+        raise NotImplementedError("delete_batch not supported for OpenSearch")
+
+
+class OpenSearchDB(DB):
+    def __init__(
+        self,
+        record_count: int,
+        dimensions: int,
+        metric: str,
+        name: str,
+        config: dict,
+    ):
+        self.host = config["opensearch_host"]
+        self.region = config["opensearch_region"]
+        self.service = "aoss"
+        self.access_key = config["aws_access_key"]
+        self.secret_key = config["aws_secret_key"]
+        self.token = config["aws_session_token"]
+
+        self.index_name = config["opensearch_index_name"]
+        self.skip_populate = config["skip_populate"]
+        self.overwrite = config["overwrite"]
+        self.dimensions = dimensions
+
+        # Create the OpenSearch client
+        awsauth = AWS4Auth(
+            self.access_key,
+            self.secret_key,
+            self.region,
+            self.service,
+            session_token=self.token,
+        )
+        self.client = OpenSearch(
+            hosts=[{"host": self.host, "port": 443}],
+            http_auth=awsauth,
+            timeout=300,
+            use_ssl=True,
+            verify_certs=True,
+            connection_class=RequestsHttpConnection,
+        )
+
+        # Create the index
+        if self.index_name is None:
+            # None specified, default to "vsb-<workload>"
+            self.index_name = f"vsb-{name}"
+
+        self.create_index()
```

Comment:
You shouldn't need to call create_index() here, it should be sufficient to just do it in `initialise_population`.